h5py/_selector.pyx

```diff
@@ -359,11 +359,12 @@ cdef class Reader:
             buf = PyArray_DATA(arr)
 
             mspace = H5Screate_simple(self.selector.rank, mshape, NULL)
+            H5Dread(self.dataset, self.h5_memory_datatype.id, mspace,
+                    self.selector.space, H5P_DEFAULT, buf)
+
         finally:
             efree(mshape)
-
-        H5Dread(self.dataset, self.h5_memory_datatype.id, mspace,
-                self.selector.space, H5P_DEFAULT, buf)
+            H5Sclose(mspace)
```

Comment:
If there's an exception before we create mspace, I think this will fail, because it will get something that's not a space ID. I think we should either have another try/finally wrapped around `H5Dread`, or initialise it to 0, and only call H5Sclose if it's non-zero.